src/Microsoft.AspNet.Server.Kestrel/Http/Frame.cs

```diff
@@ -218,33 +288,37 @@ private void FireOnCompleted()
             }
         }
 
-        private async Task ExecuteAsync()
+        public void Flush()
         {
-            Exception error = null;
-            try
-            {
-                await Application.Invoke(this).ConfigureAwait(false);
+            ProduceStart(immediate: false);
+            SocketOutput.Write(_emptyData, immediate: true);
+        }
+
+        public Task FlushAsync(CancellationToken cancellationToken)
+        {
+            ProduceStart(immediate: false);
+            return SocketOutput.WriteAsync(_emptyData, immediate: true);
+        }
+
+        public void Write(ArraySegment<byte> data)
+        {
+            ProduceStart(immediate: false);
 
-                // Trigger FireOnStarting if ProduceStart hasn't been called yet.
-                // We call it here, so it can go through our normal error handling
-                // and respond with a 500 if an OnStarting callback throws.
-                if (!_responseStarted)
+            if (_autoChunk)
+            {
+                if (data.Count == 0)
```

Comment:
I don't think this is a big deal, but it's even more subtle to say "any write, except those of zero length without the Content-Length header aleready being set, will begin the response body".